### PR TITLE
Improve client exception handling

### PR DESCRIPTION
**Pull request by janakamarasena**
Part of the fix for https://github.com/wso2/product-is/issues/7408

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 1]
```diff
@@ -423,7 +425,7 @@ public void deleteUser(String userId) throws NotFoundException, CharonException
     @Deprecated
     public List<Object> listUsersWithGET(Node rootNode, int startIndex, int count, String sortBy, String sortOrder,
                                          String domainName, Map<String, Boolean> requiredAttributes)
-            throws CharonException, NotImplementedException {
+            throws CharonException, NotImplementedException, BadRequestException {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:428]: Although `BadRequestException` the was not initially added it is available in the interface[1].

[1] https://github.com/wso2/charon/blob/master/modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/UserManager.java#L85

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 2]
```diff
@@ -437,7 +439,7 @@ public void deleteUser(String userId) throws NotFoundException, CharonException
     @Override
     public List<Object> listUsersWithGET(Node rootNode, Integer startIndex, Integer count, String sortBy,
                                          String sortOrder, String domainName, Map<String, Boolean> requiredAttributes)
-            throws CharonException, NotImplementedException {
+            throws CharonException, NotImplementedException, BadRequestException {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:442]: Although `BadRequestException` the was not initially added it is available in the interface[1].

[1] https://github.com/wso2/charon/blob/master/modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/UserManager.java#L67

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 3]
```diff
@@ -624,7 +627,19 @@ private boolean canPaginate(int offset, int limit) {
                 users.addAll(carbonUM.getUserListWithID(userIdLocalClaim, claimValue, null));
             }
             return users;
+        } catch (UserStoreClientException e) {
+            String errorMessage = String.format("Error while listing usernames from domain: %s. %s", domainName,
+                    e.getMessage());
+            throw new BadRequestException(errorMessage, ResponseCodeConstants.INVALID_VALUE);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:633]: 'e' is not logged or rethrown

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 4]
```diff
@@ -624,7 +627,19 @@ private boolean canPaginate(int offset, int limit) {
...
         } catch (UserStoreException e) {
+            // Sometimes client exceptions are wrapped in the super class.
+            // Therefore checking for possible client exception.
+            Throwable ex = ExceptionUtils.getRootCause(e);
+            if (ex instanceof UserStoreClientException) {
+                String errorMessage = String.format("Error while listing usernames from domain: %s. %s", domainName,
+                        ex.getMessage());
+                throw new BadRequestException(errorMessage, ResponseCodeConstants.INVALID_VALUE);
+            }
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:642]: 'e' is not logged or rethrown

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 3 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:636]: added

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java (hunk 4 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:648]: added

**Review by IsuraD — APPROVED** [review]

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**